### DISCUSSION TITLE
Keycloak IdP Post Login Flow - how to use it properly

### DISCUSSION THREAD

**Discussion by coderamblings** [keycloak/keycloak#14591] [Q&A] [answered]
I'm trying to set up Keycloak to restrict access to clients depending on their roles. Since the users authenticate against AAD, I'd like to use the "Post Login Flow" configured in Identitity Provider settings. Fortunately, for the moment we want to apply the same set of rules across all clients, so this option would suffice.

I've set up a very basic authentication flow Role-access1 that should check whether a user has "restrict-by-role" role assigned (copied Browser loging, removed all, added condition if that's relevant). I can see in user's "Role Mappings" that indeed the test user has the role assigned through group membership. Unfortunately, upon enabling the post login flow I get the generic "We are sorry... Invalid username or password" screens for all users. 
![Screenshot from 2022-09-23 08-44-44](https://user-images.githubusercontent.com/10697194/192362928-6df80aba-3c02-41e5-9f0a-74acffdeb5f2.png)
![Screenshot from 2022-09-23 09-06-19](https://user-images.githubusercontent.com/10697194/192362958-6ced6c9f-d882-491c-b4b3-7b09af2b163c.png)
![Screenshot from 2022-09-23 09-06-31](https://user-images.githubusercontent.com/10697194/192362978-8e8e1613-623e-4675-a646-98f5775c3f87.png)
I've checked debug logs and I can see the system attempts to execute the post-login-flow, but fails.
```
[org.keycloak.services.resources.IdentityBrokerService] (default task-6) Authorization code is valid.
[org.keycloak.services.resources.IdentityBrokerService] (default task-6) Token will not be stored for identity provider [SSO].
[org.keycloak.services.resources.IdentityBrokerService] (default task-6) Redirect to postBrokerLogin flow after authentication with identityProvider 'SSO'.
[io.undertow.request] (default I/O-3) Matched prefix path /auth for path /auth/realms/master/login-actions/post-broker-login
[io.undertow.request.security] (default task-6) Attempting to authenticate /auth/realms/master/login-actions/post-broker-login, authentication required: false
[io.undertow.request.security] (default task-6) Authentication outcome was NOT_ATTEMPTED with method io.undertow.security.impl.CachedAuthenticatedSessionMechanism@46c8c6b2 for /auth/realms/master/login-actions/post-broker-login
[io.undertow.request.security] (default task-6) Authentication result was ATTEMPTED for /auth/realms/master/login-actions/post-broker-login
[org.keycloak.transaction.JtaTransactionWrapper] (default task-6) new JtaTransactionWrapper
[org.keycloak.transaction.JtaTransactionWrapper] (default task-6) was existing? false
[org.jboss.resteasy.resteasy_jaxrs.i18n] (default task-6) RESTEASY002315: PathInfo: /realms/master/login-actions/post-broker-login
[org.keycloak.services.managers.AuthenticationSessionManager] (default task-6) Found AUTH_SESSION_ID cookie with value 7e7926d9-7018-xxxxxx
[org.keycloak.services.resources.SessionCodeChecks] (default task-6) Transition between flows! Current flow: post-broker-login, Previous flow: authenticate
[org.keycloak.authentication.AuthenticationProcessor] (default task-6) AUTHENTICATE
[org.keycloak.authentication.AuthenticationProcessor] (default task-6) AUTHENTICATE ONLY
[org.keycloak.authentication.DefaultAuthenticationFlow] (default task-6) processFlow: Role-access1
[org.hibernate.SQL] (default task-6) 
    select
        roleentity0_.ID as col_0_0_ 
    from
        KEYCLOAK_ROLE roleentity0_ 
    where
        roleentity0_.CLIENT_ROLE=false 
        and roleentity0_.NAME=? 
        and roleentity0_.REALM=?
[org.jboss.jca.core.connectionmanager.pool.strategy.OnePool] (default task-6) KeycloakDS: getConnection(null, WrappedConnectionRequestInfo@6dd2b8e0[userName=keycloak]) [0/100]
[org.hibernate.loader.Loader] (default task-6) Result set row: 0
[org.hibernate.loader.Loader] (default task-6) Result row: 
[org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (default task-6) Initiating JDBC connection release from afterStatement
[org.keycloak.authentication.DefaultAuthenticationFlow] (default task-6) check execution: 'role check flow', requirement: 'CONDITIONAL'
[org.keycloak.authentication.DefaultAuthenticationFlow] (default task-6) processFlow: role check
[org.keycloak.authentication.DefaultAuthenticationFlow] (default task-6) Flow 'role check flow' failed
[org.keycloak.services] (default task-6) KC-SERVICES0013: Failed authentication: org.keycloak.authentication.AuthenticationFlowException
        at org.keycloak.authentication.AuthenticationProcessor.authenticateOnly(AuthenticationProcessor.java:984)
        at org.keycloak.services.resources.LoginActionsService$1.authenticateOnly(LoginActionsService.java:798)
        (...)
```
Ending in:
```
[org.keycloak.events] (default task-6) type=IDENTITY_PROVIDER_POST_LOGIN_ERROR, realmId=master, clientId=https://sxxxxxxxxxxx, userId=null, ipAddress=xxxxxxxxx, error=invalid_user_credentials, identity_provider=SSO, auth_method=saml, redirect_uri=https://sxxxxxxxxxxx/oauth2/callback/saml, identity_provider_identity=xxxxx@xxxx.com, code_id=7e7926d9-7018-xxxxxxx, authSessionParentId=7e7926d9-7018-xxxxx, authSessionTabId=ILBy-xxxx
```

This is Keycloak 11.0.3. We could probably upgrade it to a more recent version if that *actually* makes a difference in this case.
Any suggestions where does the problem come from?

**Comment by coderamblings** [comment, marked as answer]
I'll write an answer for myself after getting help from a colleague.
We've since upgraded to Keycloak X, so this has been tested to work in Keycloak 19 & Keycloak 20 versions.
Same flow, pictures done if old / new console. Flow attached as post login in IdP configuration, requiring roles assigned to groups via Keycloak's admin console.

![Screenshot 2022-12-15 at 20 01 42](https://user-images.githubusercontent.com/10697194/209673695-b1385578-a97e-4327-b477-c0e57095bc53.png)
![Screenshot from 2022-12-27 14-28-08](https://user-images.githubusercontent.com/10697194/209673728-208f6c28-5ec7-40b4-8948-b69c65716daa.png)

**Reply by tymaoa2** [reply]
what a life saver, thank you !!!

**Comment by franco-viotti** [comment]
Hi, I tried your solution but still get that invalid username or password no matter what roles the user have. Any suggestions on what could I do? I'm using version 22.0
![image](https://github.com/keycloak/keycloak/assets/73352206/7f42c376-7363-4dc2-8ae6-807d2d07ea8a)

**Reply by bjconlan** [reply]
I noticed that if the flow doesn't return a success an exception is still thrown. (which results in said message... which wasnt obvious to me... might be a weird regression we are seeing in 22 (im on 22.0.4).
(in my case i had an empty idb post auth flow).

Root cause seems to be due to https://github.com/keycloak/keycloak/blob/22.0.4/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java#L229 isn't ever returning a success if nothing is run.

(as exeucted in https://github.com/keycloak/keycloak/blob/22.0.4/services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java#L986 and thrown at line 992)

I threw the 'Allow access' step in the flow (and set to required) and the flow is now executing (without the error)

**Reply by franco-viotti** [reply]
Thanks for the reply, I'll try your solution and come back to you

**Reply by tymaoa2** [reply]
@bjconlan what a life saver, thank you !!!

**Comment by afgarcia86** [comment]
OMG thank you so much for posting this!

**Comment by HolyMacarony** [comment]
Hello everyone!  so this works for exactly one role, that you can define in this config.

does anyone know if it is possible to expand this logic to:

only grant access to this generic client, when the user has one of the roles that this client has? so that it can protect every client and not just one?
